Bucket ref protection on replicas - part 1 - new fail-safe measures

Gerold103 opened the pull request:
The patchset starts work on the big issue of bucket refs being not protected from incorrect changes of _bucket. Mostly that is about the main bug which allows to garbage collect a bucket on a master instance even if it has some read-requests in progress on replicas. That makes the requests potentially access inconsistent partially deleted data.

The fix consists of multiple steps which are being sent as separate patchsets. Otherwise there would be no progress for too long while I am busy with non-vshard activities in the meantime.

#173

R-omk commented on vshard/storage/init.lua:
```diff
@@ -2685,14 +2685,21 @@ end
+        -- inconsistent data - the bucket could be half-deleted, for example.
+        -- Regardless of what user's function returns, treat it as an error.
+        return nil, err
+    end
```
I propose to write log with the original error (ret1) if it happened else it will be lost. Or write it to log in any case.

Gerold103 replied:
In the new version of the patch I added support for stacked diagnostics and return the user function's exception in `err.prev` member. The caller can now log it if he wants.

olegrok commented on vshard/storage/init.lua:
```diff
@@ -292,8 +293,62 @@ local function bucket_generation_increment()
+        end
+        return
+    end
+    -- Ignore unknown statuses. Could be set no idea how, but just ignore.
```
Maybe we can at least log them?

Gerold103 replied:
At first I thought you are right and went to add the log, but then remembered that we don't have a ratelimited log in Lua (not publicly). That means if we have a massive bucket status rename, we will produce potentially gigabytes of logs all over the cluster. I propose to keep it like that, it is at least not worse that it was. And in future commits I plan to add validation into `on_replace`, so a bad status won't even be able to commit.

olegrok reviewed (approved):
LGTM in general

I left two minor comments.

Gerold103 commented on vshard/storage/init.lua:
```diff
@@ -2625,14 +2685,21 @@ end
-    if not ok then
-        return ok, err
+    local ret1, ret2, ret3
+    local ok, ok_ref, err
```
Indeed, that was a leftover from a previous version of the patch:
```Diff
--- a/vshard/storage/init.lua
+++ b/vshard/storage/init.lua
@@ -2685,13 +2685,11 @@ end
 -- @retval nil, err Error.
 -- @retval values Success.
 local function storage_call(bucket_id, mode, name, args)
-    local ret1, ret2, ret3
-    local ok, ok_ref, err
-    ok_ref, err = bucket_ref(bucket_id, mode)
+    local ok_ref, err = bucket_ref(bucket_id, mode)
     if not ok_ref then
         return nil, err
     end
-    ok, ret1, ret2, ret3 = local_call(name, args)
+    local ok, ret1, ret2, ret3 = local_call(name, args)
     ok_ref, err = bucket_unref(bucket_id, mode)
     if not ok_ref then
         -- It should not normally happen. But the bucket could be deleted
```

Gerold103 commented:
@olegrok please, take another look. I added a new commit to support stacked diag, and reworked the last commit to return user function's error along with unref error together.

olegrok reviewed (approved):
Thanks for fixes and  changes.
